Source/ASDisplayNode.mm

```diff
@@ -829,6 +837,96 @@ - (void)nodeViewDidAddGestureRecognizer
   _flags.viewEverHadAGestureRecognizerAttached = YES;
 }
 
+- (UIEdgeInsets)fallbackSafeAreaInsets
+{
+  ASDN::MutexLocker l(__instanceLock__);
+  return _fallbackSafeAreaInsets;
+}
+
+- (void)setFallbackSafeAreaInsets:(UIEdgeInsets)insets
+{
+  {
+    ASDN::MutexLocker l(__instanceLock__);
+    ASDisplayNodeAssertThreadAffinity(self);
+
+    if (UIEdgeInsetsEqualToEdgeInsets(insets, _fallbackSafeAreaInsets)) {
+      return;
+    }
+
+    _fallbackSafeAreaInsets = insets;
+  }
+
+  BOOL needsManualUpdate = !AS_AT_LEAST_IOS11 || _flags.layerBacked;
+  BOOL updatesLayoutMargins = needsManualUpdate && self.insetsLayoutMarginsFromSafeArea;
+
+  if (needsManualUpdate) {
+    [self safeAreaInsetsDidChange];
+  }
+
+  if (updatesLayoutMargins) {
+    [self layoutMarginsDidChange];
+  }
+}
+
+- (void)_fallbackUpdateSafeAreaOnChildren
```

Comment:
I added a shortcut that skips a child if it is a view-backed node and we run on iOS 11. And `-[ASDisplayNode setFallbackSafeAreaInsets:]` returns early when the insets don't change.

Although we could try to skip calculations for the children that use fallback safe area insets but don't really need to have them updated (e.g. neither the child's frame, the parent's frame nor the parent's safe area wasn't updated), I doubt that it would save more CPU time than would spend running the checks.

The other thing I'm thinking about is that on iOS 11 we can disable fallback logic completely if we avoid supporting the safe area for the layer-backed nodes. But I'm not sure if we should do that.

Reply:
Thanks for the note. I think this implementation and your trade-offs are reasonable. Let's roll with this.